In [1]:
from tensorflow.keras.preprocessing.text import one_hot

In [2]:
### sentences
sent=[  'the glass of milk',
     'the glass of juice',
     'the cup of tea',
    'I am a good boy',
     'I am a good developer',
     'understand the meaning of words',
     'your videos are good',]

In [3]:
sent

['the glass of milk',
 'the glass of juice',
 'the cup of tea',
 'I am a good boy',
 'I am a good developer',
 'understand the meaning of words',
 'your videos are good']

In [4]:
## Define the vocabulary size
voc_size=10000

In [5]:
sent

['the glass of milk',
 'the glass of juice',
 'the cup of tea',
 'I am a good boy',
 'I am a good developer',
 'understand the meaning of words',
 'your videos are good']

In [6]:
### One Hot Representation: create a list representation
one_hot_repr=[one_hot(words,voc_size)for words in sent]
one_hot_repr

[[4356, 2252, 3703, 6076],
 [4356, 2252, 3703, 7639],
 [4356, 234, 3703, 533],
 [8608, 6014, 1101, 3340, 4435],
 [8608, 6014, 1101, 3340, 2522],
 [6476, 4356, 9019, 3703, 2721],
 [2003, 4114, 9065, 3340]]

for example for the sentence 'the glass of milk', in indices 4356, 2252, 3703, 6076, I will get value 1 but the rest of list is set to 0.

In [14]:
## word Embedding Representation

from tensorflow.keras.layers import Embedding
#from tensorflow.keras.processing.sequence import pad_sequences
from tensorflow.keras.utils import pad_sequences
from tensorflow.keras.models import Sequential

In [15]:
import numpy as np

We need to make all the sentences of the same size. So, we are using pad sequences to make them the same size.

if I print padding = 'pre' I will get all the zeros at initial indices, and if I print padding = 'post' I will get them at the last indices.

In [16]:
sent_length=8
embedded_docs=pad_sequences(one_hot_repr,padding='pre',maxlen=sent_length)
print(embedded_docs)

[[   0    0    0    0 4356 2252 3703 6076]
 [   0    0    0    0 4356 2252 3703 7639]
 [   0    0    0    0 4356  234 3703  533]
 [   0    0    0 8608 6014 1101 3340 4435]
 [   0    0    0 8608 6014 1101 3340 2522]
 [   0    0    0 6476 4356 9019 3703 2721]
 [   0    0    0    0 2003 4114 9065 3340]]


In [17]:
## feature representation
dim=10

In [20]:
model = Sequential()
model.add(Embedding(voc_size, dim, input_length = sent_length))
model.build(input_shape=(None, sent_length)) # Explicitly build the model
model.compile('adam', 'mse') # optimizer, loss function

In [21]:
model.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_2 (Embedding)         │ (None, 8, 10)          │       100,000 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 100,000 (390.62 KB)

 Trainable params: 100,000 (390.62 KB)

 Non-trainable params: 0 (0.00 B)

In [23]:
model.predict(embedded_docs)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 82ms/step


array([[[ 3.69834192e-02, -1.54219046e-02, -4.06312458e-02,
         -2.70861983e-02,  4.13918830e-02,  2.45153420e-02,
         -1.88981537e-02,  2.21059434e-02, -4.81977351e-02,
         -5.86139038e-03],
        [ 3.69834192e-02, -1.54219046e-02, -4.06312458e-02,
         -2.70861983e-02,  4.13918830e-02,  2.45153420e-02,
         -1.88981537e-02,  2.21059434e-02, -4.81977351e-02,
         -5.86139038e-03],
        [ 3.69834192e-02, -1.54219046e-02, -4.06312458e-02,
         -2.70861983e-02,  4.13918830e-02,  2.45153420e-02,
         -1.88981537e-02,  2.21059434e-02, -4.81977351e-02,
         -5.86139038e-03],
        [ 3.69834192e-02, -1.54219046e-02, -4.06312458e-02,
         -2.70861983e-02,  4.13918830e-02,  2.45153420e-02,
         -1.88981537e-02,  2.21059434e-02, -4.81977351e-02,
         -5.86139038e-03],
        [ 1.04830377e-02, -1.33503452e-02, -4.48459387e-02,
         -1.54761672e-02, -4.50728089e-03, -7.55302608e-05,
          3.78717072e-02,  5.26764244e-03,  2.461246

In [24]:
embedded_docs[0]

array([   0,    0,    0,    0, 4356, 2252, 3703, 6076], dtype=int32)

In [26]:
#model.predict(embedded_docs[0])
pred = model.predict(embedded_docs[:1])   # also gives shape (1, sequence_length)


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 73ms/step


In [27]:
pred

array([[[ 3.6983419e-02, -1.5421905e-02, -4.0631246e-02, -2.7086198e-02,
          4.1391883e-02,  2.4515342e-02, -1.8898154e-02,  2.2105943e-02,
         -4.8197735e-02, -5.8613904e-03],
        [ 3.6983419e-02, -1.5421905e-02, -4.0631246e-02, -2.7086198e-02,
          4.1391883e-02,  2.4515342e-02, -1.8898154e-02,  2.2105943e-02,
         -4.8197735e-02, -5.8613904e-03],
        [ 3.6983419e-02, -1.5421905e-02, -4.0631246e-02, -2.7086198e-02,
          4.1391883e-02,  2.4515342e-02, -1.8898154e-02,  2.2105943e-02,
         -4.8197735e-02, -5.8613904e-03],
        [ 3.6983419e-02, -1.5421905e-02, -4.0631246e-02, -2.7086198e-02,
          4.1391883e-02,  2.4515342e-02, -1.8898154e-02,  2.2105943e-02,
         -4.8197735e-02, -5.8613904e-03],
        [ 1.0483038e-02, -1.3350345e-02, -4.4845939e-02, -1.5476167e-02,
         -4.5072809e-03, -7.5530261e-05,  3.7871707e-02,  5.2676424e-03,
          2.4612460e-02, -1.8113457e-02],
        [-1.5810262e-02, -4.3406166e-02,  9.6286461e-04,  3.